File: generate_er_diagram.py
#!/usr/bin/env python3
"""
Database ER Diagram Generator
Converts database schema to JPG image
"""

import json
import subprocess
from datetime import datetime
from PIL import Image, ImageDraw, ImageFont
import os

# Database configuration
DB_CONFIG = {
    'tables': {
        'roles': {'pk': 'id', 'fk': [], 'columns': ['id', 'name', 'created_at']},
        'trades': {'pk': 'id', 'fk': [], 'columns': ['id', 'trade_name', 'description', 'created_at']},
        'users': {'pk': 'id', 'fk': ['role_id', 'trade_id'], 'columns': ['id', 'full_name', 'email', 'phone', 'password', 'role_id', 'trade_id', 'status', 'created_at']},
        'subjects': {'pk': 'id', 'fk': [], 'columns': ['id', 'name', 'description', 'created_at']},
        'study_materials': {'pk': 'id', 'fk': ['subject_id'], 'columns': ['id', 'subject_id', 'title', 'description', 'file_path', 'created_at']},
        'questions': {'pk': 'id', 'fk': [], 'columns': ['id', 'question_text', 'options', 'correct_answer', 'created_at']},
        'exams': {'pk': 'id', 'fk': [], 'columns': ['id', 'title', 'duration', 'total_questions', 'created_at']},
        'exam_questions': {'pk': 'id', 'fk': ['exam_id', 'question_id'], 'columns': ['id', 'exam_id', 'question_id', 'order']},
        'exam_attempts': {'pk': 'id', 'fk': ['user_id', 'exam_id'], 'columns': ['id', 'user_id', 'exam_id', 'start_time', 'end_time', 'score']},
        'results': {'pk': 'id', 'fk': ['user_id', 'exam_id', 'attempt_id'], 'columns': ['id', 'user_id', 'exam_id', 'attempt_id', 'score', 'created_at']},
        'certificates': {'pk': 'id', 'fk': ['user_id', 'result_id'], 'columns': ['id', 'user_id', 'result_id', 'certificate_no', 'issue_date']},
        'community_posts': {'pk': 'id', 'fk': ['user_id'], 'columns': ['id', 'user_id', 'title', 'content', 'views', 'created_at']},
        'community_comments': {'pk': 'id', 'fk': ['post_id', 'user_id'], 'columns': ['id', 'post_id', 'user_id', 'comment_text', 'created_at']},
        'otp_verifications': {'pk': 'id', 'fk': ['user_id'], 'columns': ['id', 'user_id', 'otp_code', 'expires_at', 'is_used']},
        'notifications': {'pk': 'id', 'fk': ['user_id'], 'columns': ['id', 'user_id', 'title', 'message', 'is_read', 'created_at']},
        'login_logs': {'pk': 'id', 'fk': ['user_id'], 'columns': ['id', 'user_id', 'login_time', 'ip_address', 'user_agent']},
    }
}

def create_er_diagram():
    """Create ER diagram image"""
    
    # Image dimensions
    width = 2000
    height = 2800
    
    # Create white image
    img = Image.new('RGB', (width, height), 'white')
    draw = ImageDraw.Draw(img)
    
    # Colors
    colors = {
        'darkblue': '#193366',
        'lightblue': '#add8e6',
        'white': '#ffffff',
        'black': '#000000',
        'gold': '#ffd700',
        'green': '#90ee90',
        'gray': '#f0f0f0'
    }
    
    # Try to load a nice font
    try:
        font_large = ImageFont.truetype("arial.ttf", 16)
        font_medium = ImageFont.truetype("arial.ttf", 12)
        font_small = ImageFont.truetype("arial.ttf", 10)
    except:
        font_large = ImageFont.load_default()
        font_medium = ImageFont.load_default()
        font_small = ImageFont.load_default()
    
    # Draw title
    draw.rectangle([0, 0, width, 60], fill=colors['darkblue'])
    draw.text((30, 15), "DATABASE ER DIAGRAM - exams_lms", fill=colors['white'], font=font_large)
    
    # Table positions in grid
    tables = DB_CONFIG['tables']
    col_width = 350
    row_height = 220
    start_x = 40
    start_y = 80
    cols = 5
    
    table_positions = {}
    
    # Draw tables
    idx = 0
    for table_name, table_data in tables.items():
        col = idx % cols
        row = idx // cols
        
        x = start_x + col * col_width
        y = start_y + row * row_height
        
        # Store position
        table_positions[table_name] = {'x': x, 'y': y}
        
        # Table box
        box_height = 25 + len(table_data['columns']) * 18
        draw.rectangle([x, y, x + 320, y + box_height], outline=colors['darkblue'], width=2)
        
        # Table header
        draw.rectangle([x, y, x + 320, y + 25], fill=colors['darkblue'])
        draw.text((x + 10, y + 5), table_name, fill=colors['white'], font=font_medium)
        
        # Columns
        col_y = y + 28
        for col_name in table_data['columns']:
            col_color = colors['gray']
            if col_name == table_data['pk']:
                col_color = colors['gold']
            elif col_name in table_data['fk']:
                col_color = colors['green']
            
            draw.rectangle([x+1, col_y, x+319, col_y+16], fill=col_color)
            draw.rectangle([x+1, col_y, x+319, col_y+16], outline=colors['black'])
            draw.text((x+5, col_y+1), col_name, fill=colors['black'], font=font_small)
            col_y += 17
        
        idx += 1
    
    # Footer
    footer_y = start_y + ((len(tables) // cols + 1) * row_height) + 20
    draw.text((40, footer_y), f"Total Tables: {len(tables)} | Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}", 
              fill=colors['darkblue'], font=font_small)
    
    # Legend
    legend_y = footer_y + 40
    draw.rectangle([40, legend_y, 320, legend_y+80], fill=colors['gray'], outline=colors['darkblue'])
    draw.rectangle([50, legend_y+10, 70, legend_y+25], fill=colors['gold'])
    draw.text((80, legend_y+8), "PK - Primary Key", fill=colors['black'], font=font_small)
    draw.rectangle([50, legend_y+30, 70, legend_y+45], fill=colors['green'])
    draw.text((80, legend_y+28), "FK - Foreign Key", fill=colors['black'], font=font_small)
    draw.rectangle([50, legend_y+50, 70, legend_y+65], fill=colors['lightblue'])
    draw.text((80, legend_y+48), "Column", fill=colors['black'], font=font_small)
    
    # Save
    output_dir = os.path.join(os.path.dirname(__file__), 'uploads')
    os.makedirs(output_dir, exist_ok=True)
    output_path = os.path.join(output_dir, f'ER_Diagram_{datetime.now().strftime("%Y%m%d_%H%M%S")}.jpg')
    
    img.save(output_path, 'JPEG', quality=95)
    
    return output_path

if __name__ == '__main__':
    try:
        output_file = create_er_diagram()
        print(f"✅ ER Diagram generated successfully!")
        print(f"📁 File: {output_file}")
        print(f"📊 Size: 2000x2800 pixels")
    except Exception as e:
        print(f"❌ Error: {e}")
        import traceback
        traceback.print_exc()

File: generate_er_diagram.php
<?php
/**
 * ER Diagram Generator with Image Output
 * Uses PHP GD Library to create a visual ER diagram
 */

require_once __DIR__ . '/config.php';
require_once __DIR__ . '/includes/db.php';

// Get all tables with their relationships
$tables = [];
$relationships = [];

$tables_result = $pdo->query('SELECT TABLE_NAME FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_SCHEMA = DATABASE()');
$table_names = $tables_result->fetchAll(PDO::FETCH_COLUMN);

// Collect table information
foreach ($table_names as $table) {
    $columns_result = $pdo->query("DESCRIBE $table");
    $columns = $columns_result->fetchAll(PDO::FETCH_ASSOC);
    
    $primary_key = null;
    $column_list = [];
    
    foreach ($columns as $col) {
        $col_name = $col['Field'];
        $col_type = $col['Type'];
        if ($col['Key'] === 'PRI') {
            $primary_key = $col_name;
        }
        $column_list[] = ['name' => $col_name, 'type' => $col_type, 'key' => $col['Key']];
    }
    
    // Get foreign keys
    $fk_result = $pdo->query("
        SELECT COLUMN_NAME, REFERENCED_TABLE_NAME, REFERENCED_COLUMN_NAME
        FROM INFORMATION_SCHEMA.KEY_COLUMN_USAGE
        WHERE TABLE_NAME = '$table' AND TABLE_SCHEMA = DATABASE() AND REFERENCED_TABLE_NAME IS NOT NULL
    ");
    $foreign_keys = $fk_result->fetchAll(PDO::FETCH_ASSOC);
    
    $tables[$table] = [
        'columns' => $column_list,
        'primary_key' => $primary_key,
        'foreign_keys' => $foreign_keys
    ];
    
    // Add relationships
    foreach ($foreign_keys as $fk) {
        $relationships[] = [
            'from_table' => $table,
            'from_col' => $fk['COLUMN_NAME'],
            'to_table' => $fk['REFERENCED_TABLE_NAME'],
            'to_col' => $fk['REFERENCED_COLUMN_NAME']
        ];
    }
}

// Create image
$width = 2000;
$height = 3000;
$img = imagecreatetruecolor($width, $height);

// Colors
$white = imagecolorallocate($img, 255, 255, 255);
$black = imagecolorallocate($img, 0, 0, 0);
$lightgray = imagecolorallocate($img, 240, 240, 240);
$darkblue = imagecolorallocate($img, 25, 51, 102);
$lightblue = imagecolorallocate($img, 173, 216, 230);
$gold = imagecolorallocate($img, 255, 215, 0);
$red = imagecolorallocate($img, 220, 50, 50);
$green = imagecolorallocate($img, 50, 150, 50);

// Fill background
imagefilledrectangle($img, 0, 0, $width, $height, $white);

// Title
$title = "Database ER Diagram - exams_lms";
$font_title = 5;
$title_x = ($width - strlen($title) * imagefontwidth($font_title)) / 2;
imagestringup($img, $font_title, 50, 80, $title, $darkblue);

// Position tables in a grid
$col_width = 300;
$row_height = 200;
$start_x = 50;
$start_y = 120;

$table_positions = [];
$idx = 0;
$cols = 5;

foreach ($tables as $table_name => $table_data) {
    $col = $idx % $cols;
    $row = intval($idx / $cols);
    
    $x = $start_x + $col * $col_width;
    $y = $start_y + $row * $row_height;
    
    $table_positions[$table_name] = ['x' => $x, 'y' => $y, 'width' => 280, 'height' => 30 + (count($table_data['columns']) * 18)];
    $idx++;
}

// Draw tables
foreach ($tables as $table_name => $table_data) {
    $pos = $table_positions[$table_name];
    $x = $pos['x'];
    $y = $pos['y'];
    $w = $pos['width'];
    $h = $pos['height'];
    
    // Table border
    imagerectangle($img, $x, $y, $x + $w, $y + $h, $darkblue);
    
    // Table header
    imagefilledrectangle($img, $x, $y, $x + $w, $y + 25, $darkblue);
    imagestring($img, 2, $x + 5, $y + 7, substr($table_name, 0, 25), $white);
    
    // Columns
    $col_y = $y + 28;
    foreach ($table_data['columns'] as $col) {
        $is_pk = $col['key'] === 'PRI';
        $is_fk = false;
        
        foreach ($table_data['foreign_keys'] as $fk) {
            if ($fk['COLUMN_NAME'] === $col['name']) {
                $is_fk = true;
                break;
            }
        }
        
        $col_text = '';
        if ($is_pk) $col_text = '🔑 ';
        if ($is_fk) $col_text .= '🔗 ';
        $col_text .= $col['name'];
        
        $col_color = $is_pk ? $gold : ($is_fk ? $lightblue : $white);
        imagefilledrectangle($img, $x, $col_y, $x + $w, $col_y + 16, $col_color);
        imagerectangle($img, $x, $col_y, $x + $w, $col_y + 16, $black);
        
        imagestring($img, 1, $x + 3, $col_y + 2, substr($col_text, 0, 32), $black);
        $col_y += 17;
    }
}

// Draw relationships
foreach ($relationships as $rel) {
    $from_pos = $table_positions[$rel['from_table']];
    $to_pos = $table_positions[$rel['to_table']];
    
    $from_x = $from_pos['x'] + $from_pos['width'] / 2;
    $from_y = $from_pos['y'] + $from_pos['height'];
    
    $to_x = $to_pos['x'] + $to_pos['width'] / 2;
    $to_y = $to_pos['y'];
    
    // Draw line
    imageline($img, intval($from_x), intval($from_y), intval($to_x), intval($to_y), $red);
    
    // Draw arrow
    $angle = atan2($to_y - $from_y, $to_x - $from_x);
    $arr_size = 10;
    $arrow_x1 = intval($to_x - $arr_size * cos($angle - M_PI / 6));
    $arrow_y1 = intval($to_y - $arr_size * sin($angle - M_PI / 6));
    $arrow_x2 = intval($to_x - $arr_size * cos($angle + M_PI / 6));
    $arrow_y2 = intval($to_y - $arr_size * sin($angle + M_PI / 6));
    
    imageline($img, intval($to_x), intval($to_y), $arrow_x1, $arrow_y1, $red);
    imageline($img, intval($to_x), intval($to_y), $arrow_x2, $arrow_y2, $red);
}

// Footer
$footer = "Total Tables: " . count($tables) . " | Total Relationships: " . count($relationships);
imagestring($img, 2, 50, $height - 30, $footer, $darkblue);

// Save as JPG
$output_path = __DIR__ . '/uploads/ER_Diagram.jpg';
if (!is_dir(dirname($output_path))) {
    mkdir(dirname($output_path), 0755, true);
}

imagejpeg($img, $output_path, 90);
imagedestroy($img);

echo "ER Diagram generated successfully: " . $output_path;
echo "\n\nDatabase Summary:\n";
echo "Total Tables: " . count($tables) . "\n";
echo "Total Relationships: " . count($relationships) . "\n\n";
echo "Tables:\n";
foreach ($tables as $table_name => $table_data) {
    echo "- " . $table_name . " (" . count($table_data['columns']) . " columns)\n";
}
?>
